examples/shader/extended_material.rs

```diff
@@ -0,0 +1,83 @@
+//! Demonstrates using a custom extension to the `StandardMaterial` to modify the results of the builtin pbr shader.
+
+use bevy::reflect::TypePath;
+use bevy::{
+    pbr::{ExtendedMaterial, MaterialExtension, OpaqueRendererMethod},
+    prelude::*,
+    render::render_resource::*,
+};
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_plugins(MaterialPlugin::<
+            ExtendedMaterial<StandardMaterial, MyExtension>,
```

Comment:
Given that extending materials is likely going to be very common in practice, I think its worth considering making it a "first class" aspect of the core Material trait. Using the `extended_material.rs` as an example:

```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugins(MaterialPlugin::<QuantizedStandardMaterial>::default())
        .run();
}

fn setup(
    mut commands: Commands,
    mut materials: ResMut<Assets<QuantizedStandardMaterial>>,
) {
    // sphere
    commands.spawn(MaterialMeshBundle {
        material: materials.add(QuantizedStandardMaterial {
            base: StandardMaterial {
                base_color: Color::RED,
                opaque_render_method: OpaqueRendererMethod::Auto,
                ..Default::default()
            },
            quantize_steps: 10,
        }),
        ..default()
    });
}

#[derive(Asset, AsBindGroup, TypePath, Debug, Clone)]
struct QuantizedStandardMaterial {
    base: StandardMaterial,
    #[uniform(100)]
    quantize_steps: u32,
}

impl Material for QuantizedStandardMaterial {
    type Base = StandardMaterial;
    fn fragment_shader() -> ShaderRef {
        "shaders/extended_material.wgsl".into()
    }
}

impl GetBaseMaterial<StandardMaterial> for QuantizedStandardMaterial {
    fn get_base_material(&self) -> &StandardMaterial {
        &self.base
    }
}
```

This improves a number of things ergonomically. It removes the need for the ExtendedMaterial wrapper when referencing the material type:

```rust
// This
Assets<QuantizedStandardMaterial>
// Versus this
Assets<ExtendedMaterial<StandardMaterial, QuantizedStandardMaterial>>
```
And also removes both the wrapper type and nesting when users define the material:

```rust
// This
QuantizedStandardMaterial {
    base: StandardMaterial {
        base_color: Color::RED,
        opaque_render_method: OpaqueRendererMethod::Auto,
        ..Default::default()
    },
    quantize_steps: 3,
}

// Versus this
ExtendedMaterial {
    base: StandardMaterial {
        base_color: Color::RED,
        opaque_render_method: OpaqueRendererMethod::Auto,
        ..Default::default()
    },
    extension: QuantizedStandardMaterial {
        quantize_steps: 3
    },
}
```

I've verified that this is expressible in the type system. This does have the downside of requiring this for "non-extended" materials:

```rust
impl Material for MyMaterial {
    type Base = ();
    fn fragment_shader() -> ShaderRef {
        "foo.wgsl".into()
    }
}
``` 

This UX hiccup would be resolved by "associated type defaults" (https://github.com/rust-lang/rust/issues/29661). Although progress on this feature appears to have stalled.

We _could_ also move the associated type to a normal generic:

```rust
trait Material<Base = ()> { }
```

Which enables this:

```rust
impl Material for MyMaterial {
    fn fragment_shader() -> ShaderRef {
        "foo.wgsl".into()
    }
}

impl Material<StandardMaterial> for QuantizedStandardMaterial {
    fn fragment_shader() -> ShaderRef {
        "shaders/extended_material.wgsl".into()
    }
}
``` 

But for _Rust reasons_, using a normal generic would mean that you need to re-specify the base material whenever you create a type that needs to support base materials (by explicitly naming `Material<T>`):

```rust
.add_plugins(MaterialPlugin::<QuantizedStandardMaterial, StandardMaterial>::default())
```

Reply:
Note that `GetBaseMaterial` is a separate trait (rather than existing on `Material`) in the interest of making `impl Material for MyMaterial` not need to supply a `get_base_material(&self) -> &()` method. With a separate trait, we can use a blanket impl on `T` to auto-impl this.

Reply:
I like the suggestions. I think whether we merge or not then depends on what time @robtfm has available to implement and test it.

Reply:
one issue that i see with this approach - the extension is not a `Material` but a `MaterialExtension`, and `ExtendedMaterial` works by applying `Material + MaterialExtension => Material`. The base material defines the alpha mode, opaque render method and depth bias, while the extension only defines the shaders (plus data). 

So we would need to somehow say that `E: Material` (rather than `E: MaterialExtension` is required if `Base = ()`, or change it to just use a single material trait and introduce ambiguity on the depth_bias (easily solved by adding), alpha_mode and default opaque method (not easily solved but could be done by convention).

i'll think a bit more about this later, there might be a simple way to make it fit together that i'm missing.

Reply:
One solution would be to make every "material value" return an `Option` (and default to returning None). Slightly more boilerplate when defining them, but these properties are reasonably uncommon to set manually.

Reply:
Another major downside to this strategy is that when reading a material value, you need to first check the current material, then its base, etc, etc. Given that this supports infinite nesting, we'd definitely want another set of methods to do this for us.

But supporting multiple nesting cleanly is another major benefit to this approach. The ergonomics of using `ExtendedMaterial` more than once is prohibitive:

```rust
// ExtendedMaterial
Assets<ExtendedMaterial<ExtendedMaterial<ExtendedMaterial, A>, B>, C>>
// Unified Material
Assets<C>
```

Reply:
I had the same thought about returning options. It’s not terrible.

Regarding chaining - using associated types will make the chain parent type fixed for a given extension. I guess it’s not a huge problem since the shader needs to have precise awareness of the parent anyway - at least right now, I was hoping to improve this in future though. 

The long type can be somewhat improved by using a typedef. 

But I guess overall it does seem better to use associated type. I’ll have a go

Reply:
Arg so sadly now that I've gotten into the details of this impl, I've realized that I missed an important piece:

This is a statically recursive type. For the Material trait (with `Base` associated type) alone, this is expressible. And I believe we could even make it work by checking if the TypeId is `()` and breaking the "material property resolve" recursion when we hit it (ugly but I do think it would _work_).

However this _will not work_ (and is not expressible) for the specialized pipeline key (because actual types cannot recurse like this).

Sadly that means that (with the current specialization system), I don't think we can do a unified trait.

Reply:
Too many layers would probably also hurt specialisation performance as I guess the keys would grow and grow.

Reply:
I'm going to explore one more plan that defines an ExtendedMaterial trait, but implements Material for it.

Reply:
Ok that plan hits a similar issue where we need the extending material type to implement AsBindGroup, but we need _that_ AsBindGroup impl to return the combined bind group. I think that might be possible by adding "nesting" to the AsBindGroup derive. But we're entering "scope too big to make the change now" territory. I think we should roll with the impl in this PR for now.